private/buf/bufsync/bufsyncapi/sync_handler.go

```diff
@@ -0,0 +1,526 @@
+	branchName string,
+	hash git.Hash,
+) (bool, error) {
+	repositoryID, err := h.getRepositoryID(ctx, moduleIdentity)
+	if err != nil {
+		return false, err
+	}
+	service := h.repositoryBranchServiceClientFactory(moduleIdentity.Remote())
+	res, err := service.ListRepositoryBranchesByReference(ctx, connect.NewRequest(&registryv1alpha1.ListRepositoryBranchesByReferenceRequest{
+		RepositoryId: repositoryID,
+		Reference: &registryv1alpha1.ListRepositoryBranchesByReferenceRequest_VcsCommitHash{
+			VcsCommitHash: hash.Hex(),
+		},
+	}))
+	if err != nil {
+		if connect.CodeOf(err) == connect.CodeNotFound {
+			// Repo is not created
+			return false, nil
+		}
+		return false, fmt.Errorf("list repository branch by reference: %w", err)
+	}
+	for _, branch := range res.Msg.RepositoryBranches {
+		if branch.Name == branchName {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (h *syncHandler) SyncModuleTags(
+	ctx context.Context,
+	moduleTags bufsync.ModuleTags,
+) error {
+	for _, commit := range moduleTags.TaggedCommitsToSync() {
+		repositoryID, err := h.getRepositoryID(ctx, moduleTags.TargetModuleIdentity())
+		if err != nil {
+			return err
+		}
+		referenceService := h.referenceServiceClientFactory(moduleTags.TargetModuleIdentity().Remote())
+		repositoryTagService := h.repositoryTagServiceClientFactory(moduleTags.TargetModuleIdentity().Remote())
+		commitRes, err := referenceService.GetReferenceByName(ctx, connect.NewRequest(&registryv1alpha1.GetReferenceByNameRequest{
+			Owner:          moduleTags.TargetModuleIdentity().Owner(),
+			RepositoryName: moduleTags.TargetModuleIdentity().Repository(),
+			Name:           commit.Commit().Hash().Hex(),
+		}))
+		if err != nil {
+			if connect.CodeOf(err) == connect.CodeNotFound {
+				return fmt.Errorf(
+					"git commit %q is not known to module %q",
+					commit.Commit().Hash(),
+					moduleTags.TargetModuleIdentity().IdentityString(),
+				)
+			}
+			return fmt.Errorf("get reference by name %q: %w", commit.Commit().Hash(), err)
+		}
+		if commitRes.Msg.Reference.GetVcsCommit() == nil {
+			return fmt.Errorf(
+				"git commit %q is not synced to module %q",
+				commit.Commit().Hash(),
+				moduleTags.TargetModuleIdentity().IdentityString(),
+			)
+		}
+		for _, tag := range commit.Tags() {
+			tagExists, err := h.bsrTagExists(ctx, repositoryTagService, repositoryID, tag)
+			if err != nil {
+				return fmt.Errorf("determine if tag %q exists: %w", tag, err)
+			}
+			if !tagExists {
+				_, err := repositoryTagService.CreateRepositoryTag(ctx, connect.NewRequest(&registryv1alpha1.CreateRepositoryTagRequest{
+					RepositoryId: repositoryID,
+					Name:         tag,
+					CommitName:   commitRes.Msg.Reference.GetVcsCommit().CommitName,
+				}))
+				if err != nil {
+					return fmt.Errorf("create new tag %q on module %q: %w", tag, moduleTags.TargetModuleIdentity().IdentityString(), err)
+				}
+			} else {
+				// TODO: don't do this unless we need to
```

Comment:
Don't think that's a responsibility of the CLI. The backend can check if moving the tag is a no-op.

Reply:
We can avoid the RPC if the tag already points to that commit. Why wouldn't we do that?

Reply:
Race conditions, if right after the CLI checks where the tag is, another syncer job moves the tag. But that racy behavior would still be true even if we leave that responsibility to the BSR. I guess you're right, we could skip this, as long as it doesn't mean another RPC call (to get current git hashes in the tag commit).